mp/src/game/server/momentum/mom_triggers.cpp

```diff
@@ -1578,3 +1579,345 @@ bool CNoGrenadesZone::IsInsideNoGrenadesZone(CBaseEntity *pOther)
+    DEFINE_KEYFIELD(m_flLowerThreshold, FIELD_FLOAT, "lowerThreshold"),
+    DEFINE_KEYFIELD(m_flUpperThreshold, FIELD_FLOAT, "upperThreshold"),
+    DEFINE_KEYFIELD(m_vLaunchDirection, FIELD_VECTOR, "launchDirection"),
+    DEFINE_KEYFIELD(m_target, FIELD_STRING, "launchTarget"),
+    DEFINE_KEYFIELD(m_bOnlyCheckVelocity, FIELD_INTEGER, "onlyCheckVelocity"),
+    DEFINE_OUTPUT(m_OnCatapulted, "OnCatapulted"),
+    DEFINE_KEYFIELD(m_flInterval, FIELD_FLOAT, "Interval"),
+    DEFINE_KEYFIELD(m_bOnThink, FIELD_BOOLEAN, "OnThink"),
+    DEFINE_KEYFIELD(m_flHeightOffset, FIELD_FLOAT, "heightOffset"),
+END_DATADESC()
+
+
+CTriggerMomentumCatapult::CTriggerMomentumCatapult()
+{
+    m_hLaunchTarget = nullptr;
+    m_flInterval = 1.0;
+    m_bOnThink = false;
+    m_flHeightOffset = 32.0f;
+}
+
+void CTriggerMomentumCatapult::Spawn()
+{ 
+    BaseClass::Spawn();
+
+    m_flLowerThreshold = clamp(m_flLowerThreshold, 0.0f, 1.0f);
+    m_flUpperThreshold = clamp(m_flUpperThreshold, 0.0f, 1.0f);
+
+    m_flEntryAngleTolerance = clamp(m_flEntryAngleTolerance, -1.0f, 1.0f);
+
+    if (!m_hLaunchTarget.Get())
+    {
+        if (m_target != NULL_STRING)
+        {
+            m_hLaunchTarget = gEntList.FindEntityByName(nullptr, m_target);
+            m_bUseLaunchTarget = true;
+        }
+        else
+        {
+            m_bUseLaunchTarget = false;
+        }
+    }
+}
+
+Vector CTriggerMomentumCatapult::CalculateLaunchVelocity(CBaseEntity *pOther)
+{
+    // Calculated from time ignoring grav, then compensating for gravity later
+    // From https://www.gamasutra.com/blogs/KainShin/20090515/83954/Predictive_Aim_Mathematics_for_AI_Targeting.php
+    // and setting the target's velocity vector to zero
+
+    Vector vecPlayerOrigin = pOther->GetAbsOrigin();
+
+    vecPlayerOrigin.z += m_flHeightOffset;
+
+    Vector vecAbsDifference = m_hLaunchTarget->GetAbsOrigin() - vecPlayerOrigin;
+    float flSpeed2 = m_flPlayerSpeed * m_flPlayerSpeed;
+    float flGravity = GetCurrentGravity();
+
+    float flDiscriminant = 4.0f * flSpeed2 * vecAbsDifference.Length() * vecAbsDifference.Length();
+
+    flDiscriminant = sqrtf(flDiscriminant);
+    float fTime = 0.5f * (flDiscriminant / flSpeed2);
+
+    Vector vecLaunchVelocity = (vecAbsDifference / fTime);
+
+    Vector vecGravityComp = 0.5 * (flGravity * Vector(0, 0, -1)) * fTime;
+    vecLaunchVelocity -= vecGravityComp;
+
+    return vecLaunchVelocity;
+}
+
+Vector CTriggerMomentumCatapult::CalculateLaunchVelocityExact(CBaseEntity* pOther)
+{
+    // Uses exact trig and gravity
+
+    Vector vecPlayerOrigin = pOther->GetAbsOrigin();
+
+    vecPlayerOrigin.z += m_flHeightOffset;
+
+    Vector vecAbsDifference = m_hLaunchTarget->GetAbsOrigin() - vecPlayerOrigin;
+    Vector vecAbsDifferenceXY = Vector(vecAbsDifference.x, vecAbsDifference.y, 0.0f);
+
+    float flSpeed2 = m_flPlayerSpeed * m_flPlayerSpeed;
+    float flSpeed4 = m_flPlayerSpeed * m_flPlayerSpeed * m_flPlayerSpeed * m_flPlayerSpeed;
```

Comment:
"2" and "4" are not as clear as
```suggestion
    float flSpeedSquared = m_flPlayerSpeed * m_flPlayerSpeed;
    float flSpeedQuad = m_flPlayerSpeed * m_flPlayerSpeed * m_flPlayerSpeed * m_flPlayerSpeed;
```